curvefs/src/client/s3/client_s3_cache_manager.cpp

```diff
@@ -0,0 +1,1092 @@
+                    break;
+                }
+                if ((chunkPos + len) < iter->second->GetChunkPos()) {
+                    break;
+                }
+
+                mergeDataCacheVer->emplace_back(iter->second);
+                waitDel.push_back(i);
+            }
+
+            std::vector<uint32_t>::iterator iterDel = waitDel.begin();
+            for (; iterDel != waitDel.end(); iterDel++) {
+                auto iter = dataWCacheMap_.begin();
+                std::advance(iter, *iterDel);
+                dataWCacheMap_.erase(iter);
+                s3ClientAdaptor_->GetFsCacheManager()
+                ->DataCacheNumFetchSub(1);
+            }
+            return dataCache;
+        }
+        i++;
+    }
+    return NULL;
+}
+
+DataCachePtr ChunkCacheManager::CreateWriteDataCache(
+    S3ClientAdaptorImpl* s3ClientAdaptor, uint32_t chunkPos, uint32_t len,
+    const char* data) {
+    DataCachePtr dataCache =
+        std::make_shared<DataCache>(s3ClientAdaptor, this, chunkPos, len, data);
+    LOG(INFO) << "CreateWriteDataCache chunkPos:" << chunkPos
+              << ", len:" << len;
+    WriteLockGuard writeLockGuard(rwLockWrite_);
+
+    dataWCacheMap_.emplace(chunkPos, dataCache);
+    s3ClientAdaptor_->GetFsCacheManager()->DataCacheNumInc();
+    return dataCache;
+}
+
+DataCachePtr ChunkCacheManager::CreateReadDataCache(
+    S3ClientAdaptorImpl* s3ClientAdaptor, uint32_t chunkPos, uint32_t len) {
+    DataCachePtr dataCache =
+        std::make_shared<DataCache>(s3ClientAdaptor, this, chunkPos, len);
+    LOG(INFO) << "CreateReadDataCache chunkPos:" << chunkPos << ", len:" << len;
+    std::list<DataCachePtr>::iterator iter =
+        s3ClientAdaptor->GetFsCacheManager()->Set(dataCache);
+    WriteLockGuard writeLockGuard(rwLockRead_);
+
+    dataRCacheMap_.emplace(chunkPos, iter);
+    return dataCache;
+}
+
+void ChunkCacheManager::ReleaseReadDataCache(uint64_t key) {
+    WriteLockGuard writeLockGuard(rwLockRead_);
+
+    auto iter = dataRCacheMap_.find(key);
+    if (iter == dataRCacheMap_.end()) {
+        return;
+    }
+
+    dataRCacheMap_.erase(iter);
+    return;
+}
+
+void ChunkCacheManager::ReleaseCache(S3ClientAdaptorImpl* s3ClientAdaptor) {
+    {
+        WriteLockGuard writeLockGuard(rwLockWrite_);
+        auto size = dataWCacheMap_.size();
+        dataWCacheMap_.clear();
+        if (size > 0) {
+            s3ClientAdaptor_->GetFsCacheManager()->DataCacheNumFetchSub(size);
+        }
+    }
+    WriteLockGuard writeLockGuard(rwLockRead_);
+    auto iter = dataRCacheMap_.begin();
+    for (; iter != dataRCacheMap_.end(); iter++) {
+        s3ClientAdaptor->GetFsCacheManager()->Delete(iter->second);
+    }
+    dataRCacheMap_.clear();
+}
+
+CURVEFS_ERROR ChunkCacheManager::Flush(Inode* inode, bool force) {
+    std::map<uint64_t, DataCachePtr> tmp;
+    CURVEFS_ERROR ret;
+    {
+        WriteLockGuard writeLockGuard(rwLockWrite_);
+        dataWCacheMap_.swap(tmp);
```

Comment:
I have a doubt that is it a time-consuming operation?

Reply:
dataWCacheMap_  is not too big